### DISCUSSION TITLE
external JavaFX libraries.

### DISCUSSION THREAD

**Discussion by Hugolarson** [webfx-project/webfx#17] [Q&A] [answered]
Hello,

This is a very interesting project!
I'm wondering what the limitations are, for example is it possible to ControlsFX library? Is GWT able to render these?

**Comment by salmonb** [comment, marked as answer]
Thank you @Hugolarson.

Unfortunately, most JavaFX libraries won't compile with GWT because they were not designed for that (reason 1). And even if by chance some are GWT compatible, they may use some JavaFX controls that are not yet covered by WebFX, so GWT will complain not finding these classes (reason 2).

We aim to extend the JavaFX coverage, which will reduce reason 2, but some components are complex (ex: TableView), so we will first propose simpler alternatives (ex: DataGrid) as part of our [WebFX Extras](https://github.com/webfx-extras) library. Of course this doesn't solve the problem you raised, but it will be an alternative for new WebFX applications (for example Modality will use them).

ControlsFX use many controls not yet covered, so not sure if/when we will support it. As I said, our intention is to first develop WebFX Extras rather than porting such complex JavaFX libraries.

For libraries that are *only* GWT incompatible but don't use JavaFX controls uncovered by WebFX (reason 1 but not 2), we can consider patching them to make them GWT compatible. This is the purpose of our [WebFX Libs](https://github.com/webfx-libs) space. We already ported a few. Some are also brand new (written by us).

In case you haven't read it yet, you can find a few additional info in the [limitations](https://docs.webfx.dev/#_limitations) section of the WebFX docs.
